src/sql_generator.cpp

```diff
@@ -0,0 +1,342 @@
+  auto absolute_table_name = compute_absolute_table_name(table);
+
+  sql << "UPDATE " << absolute_table_name << " SET ";
+
+  write_joined(sql, columns_regular,
+               [staging_table_name, absolute_table_name, unmodified_string](
+                   const std::string name, std::ostringstream &out) {
+                 auto colname = KeywordHelper::WriteQuoted(name, '"');
+                 out << colname << " = CASE WHEN " << staging_table_name << "."
+                     << colname << " = "
+                     << KeywordHelper::WriteQuoted(unmodified_string, '\'')
+                     << " THEN " << absolute_table_name << "." << colname
+                     << " ELSE " << staging_table_name << "." << colname
+                     << " END";
+               });
+
+  sql << " FROM " << staging_table_name << " WHERE ";
+  write_joined(
+      sql, columns_pk, [&](const std::string &pk, std::ostringstream &out) {
+        out << table.table_name << "." << KeywordHelper::WriteQuoted(pk, '"')
+            << " = " << staging_table_name << "."
+            << KeywordHelper::WriteQuoted(pk, '"');
+      });
+
+  auto query = sql.str();
+  mdlog::info("update: " + query);
+  auto result = con.Query(query);
+  if (result->HasError()) {
+    throw std::runtime_error("Could not update table <" + absolute_table_name +
+                             ">:" + result->GetError());
+  }
+}
+
+void delete_rows(duckdb::Connection &con, const table_def &table,
+                 const std::string &staging_table_name,
+                 std::vector<const column_def *> &columns_pk) {
+
+  const std::string absolute_table_name = compute_absolute_table_name(table);
+  std::ostringstream sql;
+  sql << "DELETE FROM " + absolute_table_name << " USING " << staging_table_name
+      << " WHERE ";
+
+  write_joined(
+      sql, columns_pk, [&](const std::string &pk, std::ostringstream &out) {
+        out << table.table_name << "." << KeywordHelper::WriteQuoted(pk, '"')
+            << " = " << staging_table_name << "."
+            << KeywordHelper::WriteQuoted(pk, '"');
+      });
+
+  auto query = sql.str();
+  mdlog::info("delete_rows: " + query);
+  auto result = con.Query(query);
+  if (result->HasError()) {
+    throw std::runtime_error("Error deleting rows from table <" +
+                             absolute_table_name + ">:" + result->GetError());
+  }
+}
+
+void truncate_table(duckdb::Connection &con, const table_def &table) {
+  const std::string absolute_table_name = compute_absolute_table_name(table);
+  std::ostringstream sql;
+  sql << "DELETE FROM " + absolute_table_name;
+  auto query = sql.str();
+  mdlog::info("truncate_table: " + query);
+  auto result = con.Query(query);
+  if (result->HasError()) {
+    throw std::runtime_error("Error truncating table <" + absolute_table_name +
+                             ">:" + result->GetError());
+  }
+}
+
+void check_connection(duckdb::Connection &con) {
+  auto result = con.Query("SELECT 1");
```

Comment:
This is not an adequate check for connectivity with md as it will run locally, if that is the intention.

Reply:
hmm, `PRAGMA MD_VERSION`?